### PR TITLE
copr: Add vectorize rand_with_seed

### PR DESCRIPTION
**Pull request by Rustin-Liu**
Signed-off-by: Rustin-Liu <rustin.liu@gmail.com>
PCP #5751

###  What have you changed?
Add vectorize rand_with_seed.

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?
- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

None

###  Does this PR affect `tidb-ansible`?

None

###  Refer to a related PR or issue link (optional)
#5751

###  Benchmark result if necessary (optional)
None
###  Any examples? (optional)
None

### CODE REVIEW

**Comment by Rustin-Liu** [comment]
@andylokandy  I don't know if my implement is correct, so could you please give me some suggestions? Thank you very much!

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 1]
```diff
@@ -347,6 +353,67 @@ fn pow(lhs: &Option<Real>, rhs: &Option<Real>) -> Result<Option<Real>> {
     }
 }
 
+#[inline]
+#[rpn_fn(capture = [metadata],metadata_ctor = init_rng_data)]
+fn rand(metadata: &Option<RefCell<XorShiftRng>>) -> Result<Option<Real>> {
```
Comment (andylokandy) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: I think you may not want to merge rand in this PR? Because sre-bot is stupid enough to unable to give you 100 points in single PR.

**Comment by andylokandy** [comment]
It's still incorrect. You can first try `TiDB` [here](https://play.tidb.io/) to see how corner cases of rand is handled. And you have to add those cases in unit test. I'll also provide some cases for you.

**Comment by andylokandy** [comment]
```sql
TiDB> create table test (id integer);
TiDB> insert into test values (-1);
TiDB> insert into test values (0);
TiDB> insert into test values (1);
TiDB> insert into test values (NULL);
TiDB> insert into test values (2);
TiDB> insert into test values (NULL);
TiDB> insert into test values (NULL);

TiDB> select * from test;
+-------+
| id    | 
+-------+
| -1    | 
| 0     | 
| 1     | 
| <nil> | 
| 2     | 
| <nil> | 
| <nil> | 
+-------+

TiDB> select id, rand(-1) from test;
+-------+---------------------+
| id    | rand(-1)            | 
+-------+---------------------+
| -1    | 0.3951876009960174  |     <--  Remember this number. Lets say it's X.
| 0     | 0.05964969860952545 | 
| 1     | 0.33545830491955936 | 
| <nil> | 0.5622259697326123  |
| 2     | 0.5750282403493622  | 
| <nil> | 0.3127997489885004  | 
| <nil> | 0.9342893098822316  | 
+-------+---------------------+

TiDB> select id, rand(-1) from test;
+-------+---------------------+
| id    | rand(-1)            | 
+-------+---------------------+
| -1    | 0.3951876009960174  | 
| 0     | 0.05964969860952545 |
| 1     | 0.33545830491955936 | 
| <nil> | 0.5622259697326123  | 
| 2     | 0.5750282403493622  | 
| <nil> | 0.3127997489885004  | 
| <nil> | 0.9342893098822316  | 
+-------+---------------------+

TiDB> select id, rand(0) from test;
+-------+---------------------+
| id    | rand(0)             | 
+-------+---------------------+
| -1    | 0.9451961492941164  |       <--  Also remember this number. Lets say it's Y.
| 0     | 0.24496508529377975 |
| 1     | 0.6559562651954052  | 
| <nil> | 0.05434383959970039 | 
| 2     | 0.36758720663245853 | 
| <nil> | 0.2894804331565928  | 
| <nil> | 0.19243860967493215 | 
+-------+---------------------+

TiDB> select id, rand(0) from test;
+-------+---------------------+
| id    | rand(0)             | 
+-------+---------------------+
| -1    | 0.9451961492941164  | 
| 0     | 0.24496508529377975 | 
| 1     | 0.6559562651954052  | 
| <nil> | 0.05434383959970039 | 
| 2     | 0.36758720663245853 | 
| <nil> | 0.2894804331565928  | 
| <nil> | 0.19243860967493215 | 
+-------+---------------------+

TiDB> select id, rand(NULL) from test;
+-------+---------------------+
| id    | rand(NULL)          | 
+-------+---------------------+
| -1    | 0.04556158296943171 |     <--  NULL constant seed still generates values
| 0     | 0.5360999160210664  | 
| 1     | 0.29016467874327834 | 
| <nil> | 0.8699837877508658  | 
| 2     | 0.1653830353072547  | 
| <nil> | 0.7961383143408155  | 
| <nil> | 0.1249697450147509  | 
+-------+---------------------+

TiDB> select id, rand(NULL) from test;
+-------+---------------------+
| id    | rand(NULL)          | 
+-------+---------------------+
| -1    | 0.23630693424223714 |      <--  In fact, rand_with_seed will regress to rand on NULL constant seed 
| 0     | 0.6842780534280731  | 
| 1     | 0.7371801808073687  | 
| <nil> | 0.8495363151126033  | 
| 2     | 0.8786053657057313  | 
| <nil> | 0.1857070789381935  | 
| <nil> | 0.7721479069545474  | 
+-------+---------------------+

TiDB> select id, rand(id) from test;
+-------+---------------------+
| id    | rand(id)            | 
+-------+---------------------+
| -1    | 0.3951876009960174  |     <--  It's X
| 0     | 0.9451961492941164  |     <--  It's Y
| 1     | 0.6046602879796196  | 
| <nil> | 0.08247741397717394 |     <--  NULL ColumnRef seed generates same values
| 2     | 0.16729663442585624 | 
| <nil> | 0.08247741397717394 |     <--  NULL ColumnRef seed generates same values
| <nil> | 0.08247741397717394 |     <--  NULL ColumnRef seed generates same values
+-------+---------------------+

TiDB> select id, rand(id) from test;
+-------+---------------------+
| id    | rand(id)            | 
+-------+---------------------+
| -1    | 0.3951876009960174  | 
| 0     | 0.9451961492941164  | 
| 1     | 0.6046602879796196  | 
| <nil> | 0.413938981725179   |     <--  NULL ColumnRef seed generates same values, but diffes in each request
| 2     | 0.16729663442585624 | 
| <nil> | 0.413938981725179   |
| <nil> | 0.413938981725179   |
+-------+---------------------+

```

**Comment by andylokandy** [comment]
I think the way how rand_with_seed is implemented in TiDB is guessable via the examples above.

**Comment by Rustin-Liu** [comment]
> I think the way how rand_with_seed is implemented in TiDB is guessable via the examples above.

I know what you mean. But I still confusing about how to handle ColumnRef for rand_with_seed.
I think I need to read more docs and code to understand how to handle ColumnRef.
Thank for your help.

**Comment by andylokandy** [comment]
Rand with constant seed caches the rng, while rand with ColumnRef seed don't. Instead, it creates a new rng on every call. In addition, Rand with ColumnRef seed should have a default seed for NULLs.

So you can store `(Option<RefCell<Rng>>, u64)` into metadata, where the second value is the default seed generated by time.

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 2]
```diff
@@ -389,12 +391,18 @@ fn init_rng_data(expr: &mut Expr) -> Result<Option<RefCell<XorShiftRng>>> {
     if n == 1 {
         let arg = &mut children[0];
         match arg.get_tp() {
-            ExprType::ScalarFunc => return Ok(None),
             ExprType::ColumnRef => return Ok(Some(RefCell::new(get_rng(None)))),
```
Comment (andylokandy) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: How do you distinguish it from the constant one below?

Comment (Rustin-Liu) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: I think the possible implement way:

1. if constant we cache the Rng.
2.if columRef we return Ok(None) and in rand_with_seed we use a default time seed to gen rand value.

I don't know if I am right. I have too many questions about how the cache metadata work and how to handle ColumRef. I am reading related code and articles now.

Comment (Rustin-Liu) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Any similar fn example I can ref?

Comment (andylokandy) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: @breeswish Do you have any suggestion?

Comment (andylokandy) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: After the discussion, we think you can assume that it's ColumnRef, because TiDB should not push down the constant one as it's impossible to due with distributed request. And the `rand(NULL)` behavior of `TiDB` is also incorrect while `MySQL` always returns `0.155220427694936`.

All in all, we don't have to use metadata at all in `rand_with_seed`, we just create rng by the input, and give NULLs a default seed 0, who generates `0.155220427694936`.

Comment (andylokandy) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Sorry for the confusing explanation before 😢

Comment (Rustin-Liu) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: It's ok. Thank for your help!

**Comment by Rustin-Liu** [comment]
@breeswish Can I implement rand_with_seed on #6117 ?
Because I saw in other pr the bot  grow up it can give total scores for me.

**Comment by Rustin-Liu** [comment]
working on #6117 now.
